src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -1980,7 +1980,31 @@ private void LexSyntaxTrivia(bool afterFirstToken, bool isTrailing, ref SyntaxLi
                     case '#':
                         if (_allowPreprocessorDirectives)
                         {
-                            this.LexDirectiveAndExcludedTrivia(afterFirstToken, isTrailing || !onlyWhitespaceOnLine, ref triviaList);
+                            if (isTrailing || !onlyWhitespaceOnLine)
+                            {
+                                // Directives cannot be in trailing trivia - parse as skipped tokens trivia instead.
+
+                                var directiveBuilder = new SyntaxListBuilder(10);
+                                this.LexDirectiveAndExcludedTrivia(afterFirstToken, afterNonWhitespaceOnLine: true, ref directiveBuilder);
```

Comment:
>this.LexDirectiveAndExcludedTrivia(afterFirstToken, afterNonWhitespaceOnLine: true, ref directiveBuilder); [](http://example.com/codeflow?start=32&length=106)

Is it possible that this process is going to accumulate some side information which we might not want to accumulate?

Reply:
It accumulates diagnostics attached to the tokens, but those seem useful, I think we *want* to accumulate those.

Reply:
> It accumulates diagnostics attached to the tokens

It looks like we somehow get diagnostics about missing end constructs for directives that we ignore here. Is that diagnostic reported here? Otherwise, this method has side effects beyond just tokens that it returns. I do no think we want those side effects.

Reply:
I think it would be fine to report a regular unexpected token error, skip just the `#` token, and continue parsing as if it wasn't there. What we are trying to do right now feels unnecessary complicated to me.

Reply:
> I think it would be fine to report a regular unexpected token error, skip just the `#` token, and continue parsing as if it wasn't there. What we are trying to do right now feels unnecessary complicated to me.

Sounds good, I will do that, thanks.

Reply:
> I think it would be fine to report a regular unexpected token error, skip just the `#` token, and continue parsing as if it wasn't there. What we are trying to do right now feels unnecessary complicated to me.

@AlekseyTs I have implemented this. The disadvantage of this approach is that it makes error recovery considerably worse.

To summarize the options:
- Skip `#` token, continue parsing as if it wasn't there. Worse error recovery than before.
- Parse the trailing directives into skipped tokens. Should we parse the directive as normal, including disabled text, error reporting, then just convert the tokens to skipped tokens? If not, how should we parse?
- Parse as BadDirective. A bit worse error recovery because we parse as a generic directive. Technically, it's still a directive in trailing trivia (so doesn't hold the invariant).
- Allow directives in trailing trivia in error scenarios. Very simple implementation, just remove the assert and everything continues working (error recovery, IDE refactoring features). We still have the invariant that there are no directives in trailing trivia in success scenarios. (Currently, this seems as the best option to me.)

Reply:
You can still parse it like before, but represent in the tree however you want.  This is very common in the parser. 

My main preference is simply that we do not have this structured trivia as trailing trivia.

Reply:
> I don't know how I would use it here directly because I don't have any SyntaxToken to attach the SkippedTokensTrivia onto.

Aren't you producing the token?  So you would attach to the token that you're just about to make :)

> If it's so common

By common, i mean the concept of: have the parser take steps to determine what is going on, but then determine how best to fit that into the syntactic model.

In other words, just because you parsed out a directive, doesn't mean you need to place it in the tree as a directive.

> Also Aleksey https://github.com/dotnet/roslyn/pull/75595#discussion_r1822535828 the fact we are forwarding the diagnostics from parsing the directive onto the skipped tokens

Yeah.  I don't have strong feelings/thoughts about that.  My general point is simply that i would prefer teh *final syntactic model* not have directives within trailing trivia.

--

Note: it might be easier for us to chat.  But i'll try to share my thoughts clearly here.  If you need more clarity on my views, please feel free to ask :)

--

The lexer/parser is effectively a complex program that can execute arbitrary logic to try to decide what the user code means and how it should represent it.  So, to me, it's fine for it to do things like actually parse out directives wherever it finds them, to have good error recovery.  However, that execution logic is orthogonal to the decisions it then makes in terms of final syntactic tree representation (including diagnostics).

So, *personally*, i think it would be fine to:

1. parse out as a directive, for the best recovery experience.
2. take the results of that parsing, and re-represent them as a skipped token when placing in the tree.  Note: for this i would make this extremely simple.  Literally take the entire contents of the parsed directive, and just make a single '.BadToken' whose contents are teh entirety of the text of the entire directive.

The end representation then never has a directive in trailing trivia.  Nor does it expose that we produced special directive diagnostics.  All teh user sees is a message saying `#` is not expected there.  And all the tree has is a skipped token representing that entire chunk that we don't feel is valid there.